File: lesson_12/structs_ts/src/lesson12.test.ts
import { expect, test } from '@jest/globals';
import { Lesson12 } from './lesson12.js';
import { ListNode } from './list_node.js';

function createLinkedList(values: number[]): ListNode | null {
  if (values.length === 0) return null;
  const head = new ListNode(values[0]);
  let current = head;
  for (let i = 1; i < values.length; i++) {
    current.next = new ListNode(values[i]);
    current = current.next;
  }
  return head;
}

test('testGetWinner', () => {
  const solution = new Lesson12();

  // Test case 1
  const values1 = [2, 1];
  const head1 = createLinkedList(values1);
  const expected1 = 'Even';
  expect(solution.gameResult(head1)).toBe(expected1);

  // Test case 2
  const values2 = [2, 5, 4, 7, 20, 5];
  const head2 = createLinkedList(values2);
  const expected2 = 'Odd';
  expect(solution.gameResult(head2)).toBe(expected2);

  // Test case 3
  const values3 = [4, 5, 2, 1];
  const head3 = createLinkedList(values3);
  const expected3 = 'Tie'; // Assuming "Even" is the winner in this scenario
  expect(solution.gameResult(head3)).toBe(expected3);
});

File: lesson_12/structs_java/structs_app/src/test/java/com/codedifferently/lesson12/Lesson12Test.java
package com.codedifferently.lesson12;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class Lesson12Test {

  private ListNode createLinkedList(int[] values) {
    if (values.length == 0) return null;
    ListNode head = new ListNode(values[0]);
    ListNode current = head;
    for (int i = 1; i < values.length; i++) {
      current.next = new ListNode(values[i]);
      current = current.next;
    }
    return head;
  }

  @Test
  public void testGetWinner() {
    Lesson12 solution = new Lesson12();

    // Test case 1
    int[] values1 = {2, 1};
    ListNode head1 = createLinkedList(values1);
    String expected1 = "Even";
    assertEquals(expected1, solution.gameResult(head1));

    // Test case 2
    int[] values2 = {2, 5, 4, 7, 20, 5};
    ListNode head2 = createLinkedList(values2);
    String expected2 = "Odd";
    assertEquals(expected2, solution.gameResult(head2));

    // Test case 3
    int[] values3 = {4, 5, 2, 1};
    ListNode head3 = createLinkedList(values3);
    String expected3 = "Tie"; // Assuming "Even" is the winner in this scenario
    assertEquals(expected3, solution.gameResult(head3));
  }
}
